### DISCUSSION TITLE
How to get actual router path from a request

### DISCUSSION THREAD

**Discussion by nazmul-prince** [fastapi/fastapi#12071] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
@router.put("/users/{user_id}/activate")
@inject
async def activate_user(
        user_id: int,
        db: AsyncSession = Depends(get_async_db),
        user_usecase: UserUseCaseImpl = Depends(Provide[UserUseCaseMainContainer.usecases.user_usecase])
):
    try:
        await user_usecase.activate_deactivate_user(user_id, True, db)
    except Exception as e:
        raise HTTPException(status_code=400, detail=str(e))
    return {"message": "User activated successfully"}
```

### Description

if api called like /users/1/activate then I want to get the actual router path string defined in router.get e.g. /users/{user_id}/activate
I've tried request.url.path which gives me /users/1/activate, but I want the actuall path e.g. /users/{user_id}/activate.

### Operating System

Linux

### FastAPI Version

0.112.0

### Pydantic Version

^2.8.2

### Python Version

^3.10

**Comment by AumGupta** [comment]
For the `user_id` try:

```python
from fastapi import FastAPI, Request, Depends, HTTPException
from sqlalchemy.orm import AsyncSession

app = FastAPI()

@app.put("/users/{user_id}/activate")
async def activate_user(
    user_id: int = Depends(Request.path_params.get("user_id")),
    db: AsyncSession = Depends(get_async_db),
    user_usecase: UserUseCaseImpl = Depends(Provide[UserUseCaseMainContainer.usecases.user_usecase])
):
	print(user_id)
```

**Reply by nazmul-prince** [reply]
Thanks for replying @AumGupta. actuall I'm wanting to get the route (e.g. here it will be "/users/{user_id}/activate") from my middleware what just @sinisaos mentioned below. But now the issue is I didn't able to achieve what @sinisaos suggested from inside my custom middleware.

**Comment by sinisaos** [comment]
@nazmul-prince I don't know if I understood your request correctly, but you can get the route path from request scope. Something like this. 
```python
from fastapi import FastAPI, Request

app = FastAPI()

@app.get("/users/{user_id}/activate")
async def activate_user(
    user_id: int,
    request: Request,
):
    scope_route = request.scope.get("route")
    print(scope_route.path)  # /users/{user_id}/activate
    print(user_id)  # e.g 1
```
Sorry if I missed your point.

**Reply by nazmul-prince** [reply]
Yeap thats what I'm looking for. thanks. But the issue is I want to get that scope_route (what you did above to get that) in my custom middle ware. But in the middleware I'm getting none. My middleware is as following:
```
from typing import List

import jwt
from starlette.middleware.base import BaseHTTPMiddleware, RequestResponseEndpoint
import logging

from fastapi.security import OAuth2PasswordBearer
from fastapi import HTTPException, status, Request

from infrastructure.configuration.consts.token_configs import SECRET_KEY, ALGORITHM
from infrastructure.configuration.security.role_map import ROLE_MAPS

oauth2_scheme = OAuth2PasswordBearer(tokenUrl="token")

class RBACMiddleware(BaseHTTPMiddleware):
    async def dispatch(
        self, request: Request, call_next: RequestResponseEndpoint
    ):
        scope_route = request.scope.get("route") # its giving None
        logging.info(scope_route)

        if scope_route in ROLE_MAPS:
            token = oauth2_scheme(request)

            try:
                payload = jwt.decode(token, SECRET_KEY, algorithms=ALGORITHM)
                user_roles: List[str] = payload.get("roles", [])
                required_roles = ROLE_MAPS[scope_route]

                if not any(role in user_roles for role in required_roles):
                    raise HTTPException(
                        status_code=status.HTTP_403_FORBIDDEN,
                        detail="You do not have access to this resource",
                    )
            except RuntimeError:
                raise HTTPException(
                    status_code=status.HTTP_401_UNAUTHORIZED,
                    detail="Invalid authentication credentials"
                )

        response = await call_next(request)
        return response



```
Interesting thing is while I debugged, I saw the request instance is from starlette.middleware.base._CachedRequest instead from fastapi module, donno know why? I might be doing something silly :(.

**Reply by YuriiMotov** [reply]
You can't get path in middleware since routing is performed after middlewares.


The flow of FastAPI app execution looks like this:

Middlewares -> Routing -> Endpoint


Consider using  [custom APIRoute](https://fastapi.tiangolo.com/how-to/custom-request-and-route/) instead of middleware.

**Reply by sinisaos** [reply]
@nazmul-prince As @YuriiMotov said, the middleware is called before routing. I've never used a custom `APIRoute`, but you could try something like this and see where it goes
```python
from fastapi import FastAPI, Request
from typing import Callable
from fastapi.routing import APIRoute

class RBACRouter(APIRoute):
    def get_route_handler(self) -> Callable:
        original_route_handler = super().get_route_handler()

        async def custom_route_handler(request: Request) -> Response:
            scope_route = request.scope.get("route").path
            print(scope_route) # /users/{user_id}/activate
            if scope_route in ROLE_MAPS:
                ... # rest of your code
            return await original_route_handler(request)
        return custom_route_handler

app = FastAPI()
app.router.route_class = RBACRouter

@app.put("/users/{user_id}/activate")
async def activate_user(user_id: int, request: Request):
    scope_route = request.scope.get("route")
    print(scope_route.path)  # /users/{user_id}/activate
```
If you want to stick with the middleware, you can try to iterate over `request.app.router.routes` that contains all your application routes and somehow find a way to match the desired route. Something like this
```python
class RBACMiddleware(BaseHTTPMiddleware):

    async def dispatch(self, request: Request, call_next: RequestResponseEndpoint):
        all_app_routes = request.app.router.routes # all routes in your app

```

It's not recommended, but you can try because I see people doing it online. Good luck.

**Reply by nazmul-prince** [reply]
@sinisaos thanks for you reply. Yeap I'm going for the customized api route based solution and that works, thanks again
